doc/neps/nep-0037-array-module.rst

```diff
@@ -0,0 +1,332 @@
+
+A new namespace for implicit dispatch
+=====================================
+
+Instead of supporting overrides in the main `numpy` namespace with ``__array_function__``, we could create a new opt-in namespace, e.g., ``numpy.api``, with versions of NumPy functions that support dispatching. These overrides would need new opt-in protocols, e.g., ``__array_function_api__`` patterned off of ``__array_function__``.
+
+This would resolve the biggest limitations of ``__array_function__`` by being opt-in and would also allow for unambiguously overriding functions like ``asarray``, because ``np.api.asarray`` would always mean "convert an array-like object."  But it wouldn't solve all the dispatching needs met by ``__array_module__``, and would leave us with supporting a considerably more complex protocol both for array users and implementors.
+
+We could potentially implement such a new namespace *via* the   ``__array_module__`` protocol. Certainly some users would find this convenient, because it is slightly less boilerplate. But this would leave users with a confusing choice: when should they use `get_array_module` vs. `np.api.something`. Also, we would have to add and maintain a whole new module, which is considerably more expensive than merely adding a function.
+
+Dispatching on both types and arrays instead of only types
+==========================================================
+
+Instead of supporting dispatch only via unique array types, we could also support dispatch via array objects, e.g., by passing an ``arrays`` argument as part of the ``__array_module__`` protocol. This could potentially be useful for dispatch for arrays with metadata, such provided by Dask and Pint, but would impose costs in terms of type safety and complexity.
+
+For example, a library that supports arrays on both CPUs and GPUs might decide on which device to create a new arrays from functions like ``ones`` based on input arguments:
+
+.. code:: python
+
+    class Array:
+        def __array_module__(self, types, arrays):
+            useful_arrays = tuple(a in arrays if isinstance(a, Array))
+            if not useful_arrays:
+                return NotImplemented
+            prefer_gpu = any(a.prefer_gpu for a in useful_arrays)
+            return ArrayModule(prefer_gpu)
+
+    class ArrayModule:
+        def __init__(self, prefer_gpu):
+            self.prefer_gpu = prefer_gpu
+        
+        def __getattr__(self, name):
+            import base_module
+            base_func = getattr(base_module, name)
+            return functools.partial(base_func, prefer_gpu=self.prefer_gpu)
+
+This might be useful, but it's not clear if we really need it. Pint seems to get along OK without any explicit array creation routines, and for the most part Dask is also OK with existing ``__array_function__`` style overides. Choosing whether to place an array on the CPU or GPU could be solved by making array creation lazy.
```

Comment:
I guess I don't really understand this... If the user, nowhere in their code, specified a `prefer_gpu` flag, then this wouldn't work with `__array_module__` and they'd have to use `ArrayModule` (from the example) explicitly.

`sparse` would need this to determine, for example, the `compressed_axes` in `GCXS`, and other envisioned features as well.

Reply:
Concretely, let's consider `torch.ones`: https://pytorch.org/docs/stable/torch.html#torch.ones

Currently, it is always an in memory tensor, either on CPU or GPUs. The nature is controlled either explicitly by the `device` argument or implicitly by a global "current device".

Another (hypothetical) way to handle this would be to make `torch.ones` return some sort of "lazy tensor" object (like those from dask), which only allocates memory on a particular device when the values are required, e.g., based on the devices of other arrays that are involved in the operation. This probably isn't always the best design, but in some cases it would make sense.

Reply:
I added a link to a JAX pull request that implement this: https://github.com/google/jax/pull/1668